design/014-secretstore-generator-v2.md

```diff
@@ -0,0 +1,376 @@
+
+- `providerConfig` to delegate to an out-of-tree provider.
+- `gatingPolicy` to enable/disable floodgating for generators.
+
+```yaml
+apiVersion: generator.external-secrets.io/v2alpha1
+kind: Generator
+metadata:
+  name: my-password
+  namespace: default
+spec:
+  gatingPolicy: Enabled
+  providerConfig:
+    address: http+unix:///path/to/socket.sock
+    providerRef:
+      name: password-gen
+      namespace: default
+      kind: Password
+---
+apiVersion: provider.generator.external-secrets.io/v2alpha1
+kind: Password
+metadata:
+  name: password-gen
+  namespace: default
+spec:
+  digits: 5
+  symbols: 5
+  symbolCharacters: "-_$@"
+  noUpper: false
+  allowRepeat: true
+```
+
+#### ClusterGenerator
+
+ClusterGenerators mirror ClusterSecretStores and extend namespaced Generators cluster-wide.
+
+```yaml
+apiVersion: generator.external-secrets.io/v2alpha1
+kind: ClusterGenerator
+metadata:
+  name: my-cluster-generator
+spec:
+  refreshInterval: 1m
+  controller: dev
+  providerConfig:
+    address: http+unix:///path/to/socket.sock
+    providerRef:
+      name: password-gen
+      namespace: default
+      kind: Password
+  gatingPolicy: Enabled
+  authenticationNamespace: ProviderReference # or ManifestReference
+  conditions:
+  - namespaceSelector: {}
+    namespaces: []
+    namespaceRegexes: []
+```
+
+### Changes to ExternalSecret and PushSecret
+
+To maintain compatibility, `ExternalSecret` and `PushSecret` add `secretStoreRef.apiVersion`. Controllers use this field to decide whether to call v1 providers or v2 out-of-tree providers. No other changes are required.
+
+### New Provider Interfaces
+
+Provider and Generator interfaces are updated to pass full specs and enable provider-side processing.
+
+```go
+type ProviderV2 interface {
+  GetSecret(SecretStoreSpec, ExternalSecretDataRemoteRef) ([]byte, error)
+  PushSecret(SecretStoreSpec, *corev1.Secret, PushSecretData) error
+  DeleteSecret(SecretStoreSpec, PushSecretRemoteRef) error
+  SecretExists(SecretStoreSpec, PushSecretRemoteRef) (bool, error)
+  GetAllSecrets(SecretStoreSpec, ExternalSecretFind) (map[string][]byte, error)
+  Validate(SecretStoreSpec) (admission.Warnings, error)
+  Capabilities(SecretStoreSpec) SecretStoreCapabilities
+}
+
+type GeneratorV2 interface {
+  Generate(GeneratorSpec) (map[string][]byte, GeneratorProviderState, error)
+  Cleanup(GeneratorSpec, GeneratorProviderState) error
+}
+```
+
+### Out-of-Tree Providers Maintenance
+
+#### Deployment
+
+Out-of-tree providers are separate projects with their own repos, images, and Helm charts. Users deploy ESO and the providers they need. ESO connects to providers via a Kubernetes Service indicated by `providerConfig.address`. Co-locating providers as sidecars is **discouraged** to preserve isolation and scalability.
+
+#### Governance
+
+- One repo per "official" eso maintained provider (e.g., `provider-aws`).
```

Comment:
im wondering: whats the benefit of us tracking official providers in separate repositories? Is there a specific benefit we like to see over e.g. maintaining a go.mod in a subolder? especially wondering because provider-contrib is also just one big repo in the proposal?

Reply:
Having them in a subdirectory would work for me, too. I guess i'm just used to having separate repositories. What i dislike about having everything in one repo is that in some shared areas (CI, root-level Makefile, CODEOWNERS etc) it gets quite cluttered over time.

Reply:
Mono repo makes eventual provider transfers (eg AWS SM to AWS people ) way harder. Just my own two cents)

Reply:
Or maybe you're thinking once people start adopting it, it's more difficult to point them to a new repository? I'm not too concerned about that. We could always just start deprecating it and moving to a new repo with ample of time for people to migrate out.